cpp/ql/src/experimental/Security/CWE/CWE-369/DivideByZeroUsingReturnValue.ql

```diff
@@ -0,0 +1,275 @@
+/**
+ * @name Divide by zero using return value
+ * @description Possible cases of division by zero when using the return value from functions.
+ * @kind problem
+ * @id cpp/divide-by-zero-using-return-value
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-369
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+import semmle.code.cpp.controlflow.Guards
+
+/** Holds if function `fn` can return a value equal to value `val` */
+predicate mayBeReturnValue(Function fn, Expr val) {
```

Comment:
I think this may be the source of some of your performance problems.  Rather than pairing a function with a small set of values it might return, you're pairing it with the much bigger set of every expression in the program with that value.  I suggest you change it to return the `float` directly:
```suggestion
predicate mayBeReturnValue(Function fn, float val) {
```
and adapt the places where it is used.

There may well be other performance issues in this query - I suggest you test / fix it a bit at a time rather than trying to work on the whole thing.

Reply:
thanks